File: Arrays/SubArrayCount.py
def count_subarrays_with_sum_k(nums, k):
    prefix_map = {0: 1}  # prefix_sum -> count of occurrences
    curr_sum = 0
    count = 0

    for num in nums:
        curr_sum += num

        # If (curr_sum - k) seen before, add its count to result
        if (curr_sum - k) in prefix_map:
            count += prefix_map[curr_sum - k]

        # Record current prefix sum occurrences
        prefix_map[curr_sum] = prefix_map.get(curr_sum, 0) + 1

    return count

# Example:
nums = [1, 1, 1]
k = 2
print("Number of subarrays with sum", k, ":", count_subarrays_with_sum_k(nums, k))
# Output: 2

File: Arrays/SubArrayCount.java
package Arrays;

import java.util.HashMap;

public class SubArrayCount {
    public static int subArrayCount(int[] nums, int target) {
    int n = nums.length;
    int c = 0;
    for (int i = 0; i < n; i++) {
        int sum = 0;  // Initialize sum for this start index
        for (int j = i; j < n; j++) {
            sum += nums[j];  // Add current element to sum
            if (sum == target) {
                c++;
            }
        }
    }
    return c;
}
public static int subArrayCountOptimal(int[] nums, int target) {
    int n = nums.length;
    int count = 0;
    HashMap<Integer, Integer> map = new HashMap<>();
    map.put(0, 1);
    for(int i = 0; i < n; i++){
        int c_sum = 0;
        c_sum = c_sum + nums[i];
        if(map.containsKey(c_sum - target)){
            count += map.get(c_sum - target);
        }
        map.put(c_sum, map.getOrDefault(c_sum, 0) + 1);

    }
    return count;
}

    public static void main(String[] args) {
    int[] nums = {1, 2, 3, -3, 1, 1 , 1, 4, 2, -3};
    int target = 3;
    int count = subArrayCount(nums, target);
    System.out.println("Number of subarrays with sum " + target + ": " + count);
}
}


